### ISSUE TITLE
podman image scp should allow specifying image by hash name with remote connection

### ISSUE THREAD

**Issue by Romain-Geissler-1A** [containers/podman#15762]
/kind bug

**Description**

This is a re-opening of https://github.com/containers/podman/issues/14313 which was closes but I still can't get it to work even with the very latest `quay.io/podman/upstream` image.

**Steps to reproduce the issue:**

1. Setup two remote connections named `podman1` and `podman2`

2. Build a dummy image with remote connection `podman1`:

```
[root@80525b75ca77 podman]# podman -c podman1 build .
STEP 1/1: FROM fedora
COMMIT
--> 2ecb6df9599
2ecb6df959942dd2fdeb65606ca2e42a54f8c06af10eeb594fdfc3e2656c53d1
```

3. Try to scp this image from `podman1` to `podman2`, I can't get it to work with different syntax:

```
[root@80525b75ca77 podman]# podman image scp podman1::2ecb6df959942dd2fdeb65606ca2e42a54f8c06af10eeb594fdfc3e2656c53d1 podman2::
Error: strconv.Atoi: parsing "": invalid syntax
[root@80525b75ca77 podman]# podman image scp podman1::sha256:2ecb6df959942dd2fdeb65606ca2e42a54f8c06af10eeb594fdfc3e2656c53d1 podman2::
Error: strconv.Atoi: parsing "": invalid syntax
[root@80525b75ca77 podman]# podman image scp podman1::fedora podman2::
Error: strconv.Atoi: parsing "": invalid syntax
```

**Describe the results you received:**

Image scp should work.

**Describe the results you expected:**

Argument parsing of image scp fails.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
[root@80525b75ca77 podman]# podman version
Client:       Podman Engine
Version:      4.3.0-dev
API Version:  4.3.0-dev
Go Version:   go1.18.5
Built:        Thu Jan  1 00:00:00 1970
OS/Arch:      linux/amd64```

**Comment by vrothberg** [comment]
Thanks for reaching out, @Romain-Geissler-1A!

@cdoern PTAL when you find time

**Comment by cdoern** [comment]
yep, I think I know where this is coming from. I will look for a fix today

**Comment by cdoern** [comment]
turns out this is more complicated than I thought, and needs to be brought in at the same time as the in flight ssh PRs... Will keep this issue updated as I work

**Comment by cdoern** [comment]
@Romain-Geissler-1A, fixed some scp bugs that currently exist on main. can you retry this with https://github.com/containers/podman/pull/15471 ? I cannot reproduce locally using this patch

**Comment by cdoern** [comment]
@Romain-Geissler-1A ping, any updates trying the new code?

**Comment by Romain-Geissler-1A** [comment]
Hi,

I used to use this dockerfile https://github.com/containers/podman/commit/fc95f832a597829d22b3404f9655cf38a232991c#diff-3578cf43508103a50f5808a58361f22aa6993a95eb62252f0b6bff9179fd187b to build and test pull request, but it seems you changed it for one using packages built by COPR builds. Is there any special COPR repo we can install to test pull requests ? Or the previous Dockerfile is still valid to build such a test image ?

Cheers,
Romain

**Comment by Romain-Geissler-1A** [comment]
Hi @cdoern,

I don't know if the latest image `quay.io/podman/upstream` includes you change which seems to be merged now, but I tried this image juste now, and it still doesn't work. Podman version doesn't give the build date anymore, and I don't know from which commit this podman was built.

Cheers,
Romain

**Comment by rhatdan** [comment]
@cdoern any movement on this?

**Comment by cdoern** [comment]
I'll check on this but I think this should work now.

**Comment by rhatdan** [comment]
Assuming this works.

**Comment by rhatdan** [comment]
Please open a new issue.

**Comment by rhatdan** [comment]
@cdoern PTAL

**Comment by Romain-Geissler-1A** [comment]
Hi,

I investigated this issue myself. It doesn't work for me because I actually have 2 connections and none of them are using ssh. Actually I have two connection which are targetting a local unix socket (however the real podman is potentially really remote (using either things like socat/ssh/docker exec to expose locally a unix socket which may actually target a remote podman). Looking at the code, `podman image scp` is really thought for using ssh and thus people using "regular" ssh connection or a non customized `/var/run/podman/podman.sock` socket. Now that I saw a bit more the code, I don't really understand what `podman image scp` brings as it is implemented as a `podman image save` + scp from remote machine 1 to local machine + scp from local machine to remote machine 2 + `podman image load`, and it works only in some conditions (ie not mine). What has always been doable, with either podman or docker is this one liner which isn't much more complicated (for user transfer, the same works with `sudo` which actually is the same as the podman implementation):

`podman -c connection1 image save | podman -c connection2 image load`

So I don't really get the reason why `podman images scp` exists now. I thought somehow it might avoid doing extra copies (ie try to be clever and copy only the new blobs), but apparently it's not the case (yet ?).

Anyway, I think we can consider this issue as "WONTFIX" since it was caused by the fact I misunderstood `podman image scp` works only for remote ssh connection (or local users with default connection), not any kind of generic named connection, no matter if it uses ssh://, unix:// or tcp://.

Cheers,
Romain

**Comment by rhatdan** [comment]
Well we might want it to work on these generic connections as well.  No reason to force it to be only ssh connections.

**Comment by Romain-Geissler-1A** [comment]
So this issue might be about supporting a connection like `unix:///path/to/podman.sock` (and others) which right now raises an `strconv.Atoi` error (when trying to parse `uri.getPort()` which is an empty an empty string. However since the rest of the code is all about doing ssh actions, the rest of the code has to be adapted too.

It that was me, I would avoid the file creation and the scp part, also no need to try to make a special case of the full local case vs the remote case. In all cases just fork/execing 2 podman process as above should work (passing explicitly `--url ssh://...`  or `--connection connection-name`) and linking them through a pipe.

Eventually if some "git-like" API existed in podman, allowing to list known objects on both sides and transferring just the unknown blobs between the two machines, it would be cool to cut down transfer of big images which might share common files (and this can go even further down with `zstd-chunked` and transfer only new unknown files).

**Comment by rhatdan** [comment]
SGTM, looking forward to the PRs. :^)

**Comment by lucamaf** [comment]
any idea on when this will be solved? the default podman scp command is still not working when trying to connect to a remote host using ssh
